scalene/hashablelist.py

```diff
@@ -7,49 +7,64 @@ class HashableList(list):
     This class overrides the various methods that modify the list so that they set the _hash attribute to None after modification,
     and overrides the `__hash__` method to use the SHA-1 hash of the string representation of the list.
     This way, the next time `__hash__` is invoked, it will recalculate the hash only if the list was modified.
+    The exception is the `append` and `extend` methods, which incrementally build up the hash instead of recomputing it.
 
     """
-    def __init__(self, *args):
+    def __init__(self, *args) -> None:
         super().__init__(*args)
         self._hash = None
 
-    def __hash__(self):
-        if self._hash is None:
-            self._hash = int(hashlib.sha1(str(self).encode()).hexdigest(), 16)
+    def __hash__(self) -> int:
+        if not self._hash:
+            self._hash = self._calculate_hash()
         return self._hash
 
-    def __setitem__(self, index, value):
-        super().__setitem__(index, value)
-        self._hash = None
+    def _calculate_hash(self) -> int:
+        h = hashlib.sha256()
+        for item in self:
+            h.update(str(item).encode())
+        return int(h.hexdigest(), 16)
 
-    def __delitem__(self, index):
-        super().__delitem__(index)
+    def __setitem__(self, key, value) -> None:
+        super().__setitem__(key, value)
         self._hash = None
 
-    def append(self, value):
-        super().append(value)
+    def __delitem__(self, key) -> None:
+        super().__delitem__(key)
         self._hash = None
 
-    def extend(self, iterable):
-        super().extend(iterable)
-        self._hash = None
+    def append(self, item) -> None:
+        # Ensure we have a hash value.
+        self.hash()
+        super().append(item)
+        self._hash.update(str(hash(item)).encode())
```

Comment:
This optimization is incompatible with self._hash being an integer (self._calculate_hash returns int).

self._hash can't be HASH either, because openssl hashes are not serializable.

Same bug is repeated below on new function extend() in scalene/hashatablelist.py:47.

The previous version of this function seems to works.